[libpas] Conditionally delegate large user allocations to system malloc

Achierius opened the pull request:
#### e91da8e2b02b7a89c3f3d0912083b5a678e116ac
<pre>
[libpas] Conditionally delegate large user allocations to system malloc
<a href="https://bugs.webkit.org/show_bug.cgi?id=302339">https://bugs.webkit.org/show_bug.cgi?id=302339</a>
<a href="https://rdar.apple.com/156588204">rdar://156588204</a>

Reviewed by Dan Hecht.

The system allocator (on Darwin-based platforms) possesses certain
memory-safety hardening features for which a libpas implementation
would not currently be expedient. For those allocations which
libpas is able to protect with MTE, we can rely on that feature for
equivalent protection; where that&apos;s not possible, this patch makes
it possible to fall back on the system allocator.
Such allocations are tracked as allocations in the large heap, and
are delegated ultimately to memalign/free.

Canonical link: <a href="https://commits.webkit.org/304019@main">https://commits.webkit.org/304019@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/338f2307937d1cdf4224364977453a41de337d63

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134318 "4 style errors") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/159/builds/6811 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45539 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/141881 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/86340 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7393 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6674 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/141881 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/86340 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137265 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/5157 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/120386 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/141881 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/5028 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/2642 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🛠 wpe-cairo-libwebrtc ](https://ews-build.webkit.org/#/builders/WPE-Cairo-LibWebRTC-Build-EWS "Waiting in queue, processing has not started yet") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/126389 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/114229 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38528 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144541 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/132831 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6484 "Built successfully") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39111 "Failed to checkout and rebase branch from PR 53758") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/111082 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/160/builds/6566 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/5452 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111340 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28236 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116663 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60270 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6536 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/34865 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/165779 "Built successfully") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/6369 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/43323 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/6605 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/6482 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

Achierius commented on Source/bmalloc/libpas/src/libpas/pas_mte_config.c:
```diff
@@ -142,6 +154,12 @@ static void pas_mte_do_initialization(void)
+                pas_basic_heap_runtime_config* cfg = all_bmalloc_runtime_configs[i];
+                cfg->base.max_segregated_object_size = PAS_MIN(cfg->base.max_segregated_object_size, 32 * 1024);
+                cfg->base.max_bitfit_object_size = PAS_MIN(cfg->base.max_bitfit_object_size, 32 * 1024);
+            }
```
I settled on this approach after going through a few alternatives, and while I do think it's the best way it does bump *all* bmalloc objects of sufficient size to allocate through the large heap -- including those not eligible for delegation.

The alternative that would not do this would be to instead have the segregated-directory slow-refill path check for delegation-eligibility on a per-refill basis, which would allow it to still provide marge bitfit allocations in the case that the directory knows that an allocation would not be eligible for delegation. The biggest downside with that is that it adds significantly more complexity to the question of "where and how is <whatever object> allocated".

Achierius commented on Source/bmalloc/libpas/src/libpas/pas_mte_config.c:
```diff
@@ -142,6 +154,12 @@ static void pas_mte_do_initialization(void)
+
+            for (size_t i = 0; i < sizeof(all_bmalloc_runtime_configs) / sizeof(void*); i++) {
+                pas_basic_heap_runtime_config* cfg = all_bmalloc_runtime_configs[i];
+                cfg->base.max_segregated_object_size = PAS_MIN(cfg->base.max_segregated_object_size, 32 * 1024);
```
It's worth noting that we do this (mutating runtime configs at runtime) elsewhere -- e.g. in bmalloc.cpp's enableMiniMode, which is morally doing the same thing as we are here, i.e. constraining these bounds to force certain allocations into different sub-heaps.

dhecht replied:
What is this magic 32K value? I think this is meant to use `PAS_MAX_MTE_TAGGABLE_OBJECT_SIZE`?

dhecht replied:
This does make me a bit nervous but I guess if there's precedent it should be okay.

Achierius replied:
Yes, that should be PAS_MAX_MTE_TAGGABLE_OBJECT_SIZE -- thanks

Achierius replied:
And yeah, it's used actively in production rn to constrain privileged processes to use bitfit in many cases where WCP would use segregated. Phil added it so I'm reasonably confident that it's the intended use of the heap_runtime_config

Achierius commented on Source/bmalloc/libpas/src/libpas/hotbit_heap_config.h:
```diff
@@ -80,7 +80,8 @@ PAS_API void hotbit_heap_config_activate(void);
     .marge_bitfit_page_size = PAS_MARGE_PAGE_DEFAULT_SIZE, \
-    .pgm_enabled = false)
+    .pgm_enabled = false, \
+    .delegate_large_user_allocations = true)
```
Should be false, not that it matters

Achierius replied:
After further thought it _should_ be true: if we were to switch to the hotbit heap config for whatever reason, we would not want that to disable this feature.

Achierius commented on Source/bmalloc/libpas/src/libpas/pas_large_heap.c:
```diff
@@ -172,15 +190,43 @@ pas_large_heap_try_allocate(pas_large_heap* heap,
+            return result;
+        }
+        result.zero_mode = pas_zero_mode_may_have_non_zero;
+        result.did_succeed = true;
```
looking at this block I might pull it out into a delegate_allocate_impl function

Achierius commented on Source/bmalloc/libpas/src/libpas/pas_large_heap.c:
```diff
@@ -155,6 +159,20 @@ static pas_allocation_result allocate_impl(pas_large_heap* heap,
+    return (pas_system_heap_is_enabled(heap_config->kind)
+            && heap_config->delegate_large_user_allocations
+            && heap_config->mmap_capability == pas_may_mmap
+            && allocation_mode == pas_non_compact_allocation_mode);
```
Strictly not necessary but bounds the impact radius to begin with

Achierius commented on Source/bmalloc/libpas/src/libpas/pas_large_heap.c:
```diff
@@ -155,6 +159,20 @@ static pas_allocation_result allocate_impl(pas_large_heap* heap,
+                                                             const pas_heap_config* heap_config)
+{
+    if (!PAS_MTE_IS_HARDENED)
+        return false;
```
If performance permits we will also deploy this for non-privileged processes, but this will be the initial config

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_large_heap.c:
```diff
@@ -172,13 +222,21 @@ pas_large_heap_try_allocate(pas_large_heap* heap,
+                            pas_physical_memory_transaction* transaction,
+                            bool user_facing_allocation)
 {
+    PAS_UNUSED_PARAM(user_facing_allocation);
```
Is this still needed?  I see that `user_facing_allocation` is used unconditionally below.  So, looks like this is stale.

Achierius replied:
Removed

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_large_heap.c:
```diff
@@ -155,6 +159,52 @@ static pas_allocation_result allocate_impl(pas_large_heap* heap,
+        break;
+    case pas_maybe_compact_allocation_mode:
+        PAS_ASSERT_NOT_REACHED();
+        break;
```
given the logic in `should_delegate_user_allocation_to_system_malloc` are these cases ever taken?

Achierius replied:
`maybe_compact` is not thus the NOT_REACHED, but you're right about `pas_always_compact_allocation_mode`. I left that in because in principle it should be possible, and it's definitely desirable, to protect memory that couldn't be mte-tagged -- I just disabled that at the level of `should_delegate_user_allocation_to_system_malloc` for the initial release to minimize the blast-radius. But if things go smoothly I would want to enable it later on for better coverage.

dhecht replied:
I think it's best to not add dead (i.e. untested code) until it's exercised, so I think it's best to leave the `pas_always_compact_allocation_mode` case as `NOT_REACHED` as well and just add that code in the PR that enables it.

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_large_heap.c:
```diff
@@ -155,6 +159,52 @@ static pas_allocation_result allocate_impl(pas_large_heap* heap,
+    PAS_TESTING_ASSERT(pas_system_heap_is_enabled(heap_config->kind));
+    switch (allocation_mode) {
+    case pas_non_compact_allocation_mode:
+        result.begin = (uintptr_t)pas_system_heap_memalign(alignment ? alignment : 1, size);
```
`malloc.h` says this:

```
extern void * __sized_by_or_null(size) malloc_zone_memalign(malloc_zone_t *zone, size_t alignment, size_t size) __alloc_align(2) __alloc_size(3) _MALLOC_TYPED(malloc_type_zone_memalign, 3) __OSX_AVAILABLE_STARTING(__MAC_10_6, __IPHONE_3_0);
	/*
	 * Allocates a new pointer of size size whose address is an exact multiple of alignment.
	 * alignment must be a power of two and at least as large as sizeof(void *).
	 * zone must be non-NULL.
	 */
```

so I think passing alignment of 1 is not allowed.

Also, I was wondering if there might be some disadvantages to calling memalign rather than malloc in cases where alignment is not requested.

Achierius replied:
Yeah this is an artifact of when I was enforcing some higher alignment that it turned out is not actually necessary. Changed.

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_large_heap.c:
```diff
@@ -155,6 +159,52 @@ static pas_allocation_result allocate_impl(pas_large_heap* heap,
+        result.did_succeed = false;
+        return result;
+    }
+    result.zero_mode = pas_zero_mode_may_have_non_zero;
```
what does this do? I guess the caller takes care of zeroing when necessary if this is set?

Achierius replied:
Yep exactly. If it's `pas_zero_mode_all_zeroes` then this lets the caller skip that step. E.g. if the memory was freshly mmapped then this avoids some overhead.

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_large_heap.c:
```diff
@@ -155,6 +159,52 @@ static pas_allocation_result allocate_impl(pas_large_heap* heap,
+        return false;
+    return (pas_system_heap_is_enabled(heap_config->kind)
+            && heap_config->delegate_large_user_allocations
+            && heap_config->mmap_capability == pas_may_mmap
```
This combination seems like it's easy to create configs that silently ignores `delegate_large_user_allocations`. Maybe we should instead disallow that config combination?

Also, why is `pas_may_mmap` required for delegation?

Achierius replied:
That's a good point, I'll find a place to check for and exclude that combo.

> why is pas_may_mmap required for delegation?

Because when it's set to false, that means that the user passed already-mapped memory into libpas for the allocator to use -- e.g. the JIT allocator. In that case the user depends on the memory being in a certain region, so we can't go and allocate out of who-knows-where instead.

dhecht commented on Source/bmalloc/libpas/src/libpas/bmalloc_heap_config.h:
```diff
@@ -78,7 +78,8 @@ PAS_API void bmalloc_heap_config_activate(void);
     .marge_bitfit_page_size = PAS_MARGE_PAGE_DEFAULT_SIZE, \
-    .pgm_enabled = true)
+    .pgm_enabled = true, \
+    .delegate_large_user_allocations = true)
```
Can you remind me in which cases each of these heap configs are used?

Achierius replied:
My understanding is that in practice we use three heap configs in production:
 - bmalloc covers FastMalloc, TZoneMalloc, and the gigacage
 - jit is just for the JIT heap
 - utility is for libpas-internal stuff
Then there's also the thingy-heap which is used for testing only

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_large_heap.h:
```diff
@@ -64,7 +64,8 @@ pas_large_heap_try_allocate(pas_large_heap* heap,
                             const pas_heap_config* config,
-                            pas_physical_memory_transaction* transaction);
+                            pas_physical_memory_transaction* transaction,
+                            bool use_system_malloc);
```
at the definition this is called `user_facing_allocation`.

Achierius replied:
Removed this argument entirely

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_local_allocator_inlines.h:
```diff
@@ -1737,8 +1737,6 @@ pas_local_allocator_try_allocate_slow_impl(pas_local_allocator* allocator,
                     pas_local_allocator_config_kind_get_string(allocator->config_kind));
         }
 
-        PAS_TESTING_ASSERT(allocator->config_kind != pas_local_allocator_config_kind_unselected);
```
why is this no longer true?

Achierius replied:
Duplicate of line 1742/1740 -- not related to the patch, mb

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_mte_config.h:
```diff
@@ -180,6 +180,9 @@ extern Slot g_config[];
 #define PAS_MTE_FEATURE_ENABLED(feature) (0)
+#define PAS_MTE_MEDIUM_TAGGING_ENABLED (0)
+#define PAS_MTE_IS_LOCKDOWN_MODE (0)
+#define PAS_MTE_IS_HARDENED (0)
```
Why are these now needed? How did it work before?

Achierius replied:
Previously any sites outside of pas_mte.h &co that wanted to access these would do so through PAS_MTE_HANDLE or otherwise explicitly guard for PAS_USE_MTE, but doing so here would have polluted a lot of code with ifdefs or added more macro-indirection, both of which I want to avoid.

Achierius replied:
Nevertheless I went in and replaced it with a new PAS_MTE_ENABLE_LARGE_OBJECT_DELEGATION macro so that it would be more clear what was going on

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_large_heap.c:
```diff
@@ -172,13 +222,21 @@ pas_large_heap_try_allocate(pas_large_heap* heap,
                             const pas_heap_config* heap_config,
-                            pas_physical_memory_transaction* transaction)
+                            pas_physical_memory_transaction* transaction,
+                            bool user_facing_allocation)
```
Where is this function called with `false`? If it's not, why bother having this extra complexity?

Achierius replied:
I wanted to avoid the case where someone used pas_large_heap_allocate for something where they'd be surprised to get a system-malloc allocation back. Currently we only do that via pas_large_heap_allocate_and_forget so it's possible to just have that be a semantic difference between the two functions, so I renamed pas_large_heap_allocate to specify that it's only for user-allocations.

dhecht replied:
This is better, thanks.

dhecht commented on Source/bmalloc/libpas/src/libpas/thingy_heap_config.h:
```diff
@@ -78,7 +78,8 @@ PAS_BEGIN_EXTERN_C;
     .marge_bitfit_page_size = PAS_MARGE_PAGE_DEFAULT_SIZE, \
-    .pgm_enabled = false)
+    .pgm_enabled = false, \
+    .delegate_large_user_allocations = false)
```
Is this the heap config used by pas_test? 
Does pas_test exercise this new code? If not, can it?

Achierius replied:
Some pas-tests use the thingy heap, but those expect allocations to be returned from libpas itself so that they can do some poking and such

However pas-tests also leverage the bmalloc-heap in various places, which is how this is tested right now. I've also added a new test to the BmallocTests.cpp which will force the size-limits downwards so that we get coverage sizes between 32K and the normal large-heap allocation cutoff, which was not happening previously.

Achierius replied:
That required some changes elsewhere in the codebase though since my toolchain started complaining about some existing casts in pas_mte.h only after I added some includes (e.g. pas_internal_config.h) into BmallocTests.cpp

Achierius replied:
Though I should clarify: the one thing this can't test is (unfortunately) actually allocating things from the system heap, because the system heap is not available in the pas-tests-harness since it doesn't link Bmalloc. From my initial investigations before Thanksgiving I believe that changing that will require refactoring the SystemHeap logic itself, which I'm averse to doing at this point in the cycle given the stability risks

dhecht replied:
Okay. I think worth filing a test bug to follow up.

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_large_heap.c:
```diff
@@ -155,6 +159,57 @@ static pas_allocation_result allocate_impl(pas_large_heap* heap,
+    PAS_TESTING_ASSERT(pas_system_heap_is_enabled(heap_config->kind));
+    switch (allocation_mode) {
+    case pas_non_compact_allocation_mode:
+        if (alignment > 1)
```
given the malloc.h comment, maybe we should use `> sizeof(void*)`

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_large_heap.c:
```diff
@@ -155,6 +159,57 @@ static pas_allocation_result allocate_impl(pas_large_heap* heap,
+        else
+            result.begin = (uintptr_t)pas_system_heap_malloc(size);
+        break;
+    case pas_always_compact_allocation_mode:
```
I think it's best to not add dead (i.e. untestable) code, so I'd leave this as `NOT_REACHED` and add this case in when you enable delegation for `pas_always_compact_allocation_mode`.

Achierius replied:
Ok, I'll do so

dhecht commented on Source/bmalloc/libpas/src/test/TestHarness.cpp:
```diff
@@ -717,40 +717,7 @@ int main(int argc, char** argv)
-    ADD_SUITE(TLCDecommit);
-    ADD_SUITE(TSD);
-    ADD_SUITE(Utils);
-    ADD_SUITE(ViewCache);
```
I think you mean to revert this

Achierius replied:
Just saw that while re-reading myself :')

Sorry about that!

dhecht reviewed (approved):
r=me with comments

Achierius committed:
[libpas] Conditionally delegate large user allocations to system malloc
https://bugs.webkit.org/show_bug.cgi?id=302339
rdar://156588204

Reviewed by Dan Hecht.

The system allocator (on Darwin-based platforms) possesses certain
memory-safety hardening features for which a libpas implementation
would not currently be expedient. For those allocations which
libpas is able to protect with MTE, we can rely on that feature for
equivalent protection; where that's not possible, this patch makes
it possible to fall back on the system allocator.
Such allocations are tracked as allocations in the large heap, and
are delegated ultimately to memalign/free.

Canonical link: https://commits.webkit.org/304019@main
